website/blog/2024/09-30-jsdom-based-scraping/index.md

```diff
@@ -0,0 +1,262 @@
+
+Before making calls to this API, we will need few required headers (auth data, so we will first make the call to `https://ads.tiktok.com/business/creativecenter/inspiration/popular/hashtag/pad/en`
+We will start this approach by creating a function that will create the URL for the API call for us and, make the call and get the data.
+
+```js
+export const createStartUrls = (input) => {
+    const {
+        days = '7',
+        country = '',
+        resultsLimit = 100,
+        industry = '',
+        isNewToTop100,
+ } = input;
+
+    const filterBy = isNewToTop100 ? 'new_on_board' : '';
+    return [
+ {
+            url: `https://ads.tiktok.com/creative_radar_api/v1/popular_trend/hashtag/list?page=1&limit=50&period=${days}&country_code=${country}&filter_by=${filterBy}&sort_by=popular&industry_id=${industry}`,
+            headers: {
+                // required headers
+ },
+            userData: { resultsLimit },
+ },
+ ];
+};
+```
+
+In the above function, we create the start url for the API call that include various parameters as we talked about earlier. After creating the URL according to the parameters it will call the `creative_radar_api` and fetch all the results.
+
+But it won’t work until we get the headers. So, let’s create a function that will first create a session using sessionPool and proxyConfiguration.
+
+```js
+export const createSessionFunction = async (
+    sessionPool,
+    proxyConfiguration,
+) => {
+    const proxyUrl = await proxyConfiguration.newUrl(Math.random().toString());
+    const url =
+        'https://ads.tiktok.com/business/creativecenter/inspiration/popular/hashtag/pad/en';
+    // need url with data to generate token
+    const response = await gotScraping({ url, proxyUrl });
+    const headers = await getApiUrlWithVerificationToken(
+        response.body.toString(),
+        url,
+ );
+    if (!headers) {
+        throw new Error(`Token generation blocked`);
+ }
+    log.info(`Generated API verification headers`, Object.values(headers));
+    return new Session({
+        userData: {
+            headers,
+ },
+        sessionPool,
+ });
+};
+```
+
+In this function, the main goal is to call `https://ads.tiktok.com/business/creativecenter/inspiration/popular/hashtag/pad/en` and get headers in return. To get the headers we are using `getApiUrlWithVerificationToken` function.
+
+:::note
+
+Before going ahead, I want to mention that Crawlee natively supports JSDOM using the [JSDOM Crawler](https://crawlee.dev/api/jsdom-crawler). It gives a framework for the parallel crawling of web pages using plain HTTP requests and jsdom DOM implementation. It uses raw HTTP requests to download web pages, it is very fast and efficient on data bandwidth.
+
+:::
+
+Let’s see how we are going to create the `getApiUrlWithVerificationToken` function:
+
+```js
+const getApiUrlWithVerificationToken = async (body, url) => {
```

Comment:
and this is pretty much the same, the formatting is completely off on many places